### ISSUE TITLE
Starting field line is not computed for multiple flux surfaces

### ISSUE THREAD

**Issue by krystophny** [itpplasma/SIMPLE#33]
Right now, `init_starting_surf` is only called as if there was only a single `sbeg`. The problem was masked, since `integrate_mfl_can` is defined outside a module, and it eats an array argument that was a scalar before without warning.

Solution: Do the computation of `init_starting_surf` to all flux surfaces inside `sbeg` array and change affected variables such as `volstart` to arrays.

**Comment by krystophny** [comment]
*Automated triage pass (Claude), grounded in the repository at `d8b0e01`. Verify before acting.*

Still reproducible on `main`. The masking mechanism described in 2023 is gone (`integrate_mfl_can` now lives in `module alpha_lifetime_sub`, `src/sub_alpha_lifetime_can.f90:1`, and is imported with `only:` at `src/samplers.f90:24`), but the actual defect — only `sbeg(1)` is ever traced — is unchanged, and the TODO from this issue is still sitting in the source at `src/params.f90:36`.

## Purpose

`startmode = 1` is SIMPLE's primary marker initialisation: a field line is followed on a flux surface, and markers are drawn along it with weight proportional to the flux-tube volume element, giving a volume-uniform distribution on that surface. Loss studies routinely want several surfaces at once (a radial scan of confined fraction in one run, or a radially spread population feeding the classifier). The namelist advertises that capability — the repo-root `simple.in:12-13` ships `sbeg = 0.6d0, 0.7d0` with `num_surf = 2` and the comment "The particles will be distributed equally along these flux surfaces" — but the code silently ignores `sbeg(2:)`. Anyone running the shipped default input gets a single-surface result while the input file says two. That is a silent physics error, not a missing feature.

## Current state

`init_starting_surf` traces exactly one field line, from `sbeg(1)`, and does not reference `num_surf` at all (`src/samplers.f90:23-52`):

```fortran
    subroutine init_starting_surf
        use alpha_lifetime_sub, only: integrate_mfl_can
        use params, only: dphi, nper, npoiper, phibeg, thetabeg, volstart, &
                          xstart, sbeg, bmin, bmax, bmod00

        integer :: ierr = 0
        real(dp), dimension(npoiper*nper) :: bstart     ! LOCAL, shadows params::bstart
        ...
        call integrate_mfl_can( &
            npoiper*nper, dphi, sbeg(1), phibeg, thetabeg, &
            xstart, bstart, volstart, bmod00, ierr)
        ...
        bmax = maxval(bstart)
        bmin = minval(bstart)
```

Note the shadow: `bstart` here is a **local** automatic array (`src/samplers.f90:29`). The module-level `bstart` declared at `src/params.f90:35` and allocated at `src/params.f90:500` is never written by the production path — only `test/tests/test_coord_trans.f90` uses a `bstart` of its own. Any fix has to decide deliberately whether per-surface `bstart` becomes module state or stays local; today's code is inconsistent about it.

The module storage is rank-deficient (`src/params.f90:35-37`, `src/params.f90:500`):

```fortran
    real(dp), dimension(:), allocatable :: bstart, volstart
    !where npoi is used, add a dimension at the end for sbeg
    real(dp), dimension(:, :), allocatable :: xstart
...
        allocate (xstart(3, npoi), bstart(npoi), volstart(npoi))
```

The consumer is single-surface too — `sample_surface_fieldline_impl` does one `binsrc` over the flat `volstart` for every particle (`src/samplers.f90:261-287`):

```fortran
        use params, only: volstart, ibins, xstart, npoiper, nper
        ...
            call binsrc(volstart, 1, npoiper*nper, xi, i)
            ibins = i
```

`num_surf` today only gates downstream machinery, never the field-line producer: it selects the volume-sampler bracket for `startmode == 5` (`src/simple_main.f90:862-870`), decides whether the `bminmax` cache is built (`needs_bminmax_cache`, `src/simple_main.f90:942-950`), and switches classification to per-particle `get_bminmax` (`src/classification.f90:158`, `src/simple_main.f90:2128`). So the multi-surface *consumers* already exist; only the *producer* is missing.

Two adjacent defects worth fixing in the same PR:

- `integrate_mfl_can` declares `integer, intent(in) :: npoi, ierr` (`src/sub_alpha_lifetime_can.f90:285`) yet passes `ierr` to `chamb_can` as its error output (`src/sub_alpha_lifetime_can.f90:300` and `:319`). `chamb_can` has no `intent` on its dummy (`src/chamb_m.f90:22`) so this compiles, but it is nonconforming: the caller's `ierr` is not guaranteed to receive the value, so the "points outside the chamber" guard at `src/samplers.f90:42-43` is not reliably reached. Separately, `integer :: ierr = 0` at `src/samplers.f90:28` is an initialised local, which implies `SAVE` — it is initialised once, not per call.
- `test_samplers.x` is built but never registered as a ctest (`test/tests/CMakeLists.txt:408-409` has `add_executable`/`target_link_libraries` and no `add_test`), and `test_sample_surface` (`test/tests/test_samplers.f90:50-71`) is dead: it prints `"test_sample_surface not implemented"` and returns at lines 58-59, before any of its real body runs. The fixture `test/test_data/simple.multiple_fieldline.in` (`sbeg = 0.5d0, 0.6d0`) is referenced by nothing in the tree (`rg multiple_fieldline` returns only the file itself).

## The `num_surf = 0` trap — read this before writing the loop

`startmode` defaults to `1` (`src/params.f90:57`) and `num_surf` defaults to `1` (`src/params.f90:136`), but `num_surf = 0` is a *supported* value meaning "distribute in volume", and `init_starting_surf` is called **unconditionally** for both the chartmap and VMEC paths (`src/simple_main.f90:196`, `src/simple_main.f90:207`) regardless of `startmode` or `num_surf` — commit `c7ca093` ("Always init starting surface for bmin/bmax") made that deliberate, because `bmin`/`bmax` are taken from the traced line.

Two golden-record cases hit exactly this combination:

- `test/golden_record/classifier_fast/simple.in` — `num_surf = 0`, `sbeg = 0.2`, **no `startmode` key** → `startmode = 1`
- `test/golden_record/classifier_combined/simple.in` — same

A naive `do isurf = 1, num_surf` therefore traces **zero** field lines for those two runs, leaves `xstart`/`volstart` zeroed, and makes `bmin = bmax = 0`. That silently destroys both goldens.

The correct shape is a locally derived count that never mutates the namelist value:

```fortran
n_surf_traced = max(num_surf, 1)   ! num_surf == 0 means "volume sampling", still needs one line for bmin/bmax
```

Do **not** clamp `params::num_surf` itself. `needs_bminmax_cache` reads `num_surf /= 1` (`src/simple_main.f90:948`), which is `.true.` for `num_surf = 0` today and would flip to `.false.` if `num_surf` were rewritten to `1`; the `startmode == 5` branch at `src/simple_main.f90:862` also tests `0 == num_surf` explicitly. Both would break.

## Reproduction / evidence
VMEC-free, using the CMake-generated chartmap fixture that `test_chartmap_startmode1` already relies on (`test/tests/CMakeLists.txt:534-542`). `simple.x` takes the config path as argument 1 and otherwise defaults to `simple.in` in the cwd (`src/simple_main.f90:135-137`).

```bash
cd <repo>
make CONFIG=Fast
cmake --build build --target generate_test_boozer_chartmap

mkdir -p /tmp/simple33 && cd /tmp/simple33
cp <repo>/build/test/tests/test_boozer_chartmap.nc .
cat > simple.in <<'EOF'
&config
trace_time = 1d-6
ntimstep = 4
ntestpart = 64
num_surf = 2
sbeg = 0.3d0, 0.7d0
field_input = 'test_boozer_chartmap.nc'
coord_input = 'test_boozer_chartmap.nc'
isw_field_type = 2
startmode = 1
integmode = 1
relerr = 1d-10
deterministic = .True.
generate_start_only = .True.
/
EOF
<repo>/build/simple.x simple.in
awk '{printf "%.4f\n", $1}' start.dat | sort -u
```

**Predicted from the code** (not executed in this triage pass): column 1 of `start.dat` contains a single narrow cluster around the surface traced from `sbeg(1) = 0.3`; `sbeg(2) = 0.7` never appears, because `src/samplers.f90:38-40` passes `sbeg(1)` and nothing else. Expected after the fix: two clusters, roughly `ntestpart/num_surf = 32` markers each. Run the command above to confirm the "before" state before starting work.

## Behaviour (BDD)

```gherkin
Scenario: two starting surfaces produce two marker populations
  Given simple.in with startmode = 1, num_surf = 2, sbeg = 0.3d0, 0.7d0,
        ntestpart = 64, field_input = 'test_boozer_chartmap.nc'
  When ./build/simple.x simple.in is run with generate_start_only = .True.
  Then start.dat contains 32 markers whose column 1 is within 1d-6 of the
       surface traced from s = 0.3 and 32 within 1d-6 of the one traced from
       s = 0.7

Scenario: num_surf = 0 still traces one field line for bmin/bmax
  Given startmode = 1 and num_surf = 0 and sbeg = 0.2
  When init_starting_surf returns
  Then exactly one field line has been traced from sbeg(1) = 0.2,
       params::bmin and params::bmax are nonzero and equal to today's values,
       and params::num_surf is still 0 when needs_bminmax_cache() is evaluated

Scenario: single surface behaviour is bit-identical to today
  Given test/golden_record/canonical/simple.in (sbeg = 0.3d0, no num_surf key)
  When make test-golden is run
  Then confined_fraction.dat and times_lost.dat match the golden records
       within the existing tolerances of test/golden_record/compare_files.py

Scenario: bmin/bmax span all requested surfaces
  Given num_surf = 2 and sbeg = 0.3d0, 0.7d0
  When init_starting_surf returns
  Then params::bmin equals the minimum over both traced field lines and
       params::bmax the maximum

Scenario: inconsistent num_surf is rejected, not silently truncated
  Given startmode = 1 and num_surf = 1025 (beyond size(sbeg) = 1024)
  When ./build/simple.x simple.in is run
  Then it stops with a message naming num_surf and sbeg rather than reading
       past the end of sbeg
```

## Implementation plan

1. `src/params.f90` — give the field-line arrays a trailing surface dimension, as the existing comment at line 36 instructs. Change the declarations at lines 35-37 to `real(dp), dimension(:,:), allocatable :: bstart, volstart` and `real(dp), dimension(:,:,:), allocatable :: xstart`. In `reallocate_arrays` (line 467; the current `allocate` is line 500) allocate `xstart(3, npoi, n_surf)`, `bstart(npoi, n_surf)`, `volstart(npoi, n_surf)` with `n_surf = max(num_surf, 1)` — a local, do not write back to `num_surf`. Also `error stop` there if `num_surf > size(sbeg)` (`sbeg` is `dimension(1024)`, `src/params.f90:33`).
2. `src/samplers.f90:23-52` — rewrite `init_starting_surf` as a loop over `isurf = 1, max(num_surf, 1)` calling `integrate_mfl_can(npoiper*nper, dphi, sbeg(isurf), phibeg, thetabeg, xstart(:,:,isurf), bstart(:,isurf), volstart(:,isurf), bmod00_surf, ierr)`. **Delete the local `bstart` declaration at line 29** and use `params::bstart` instead, otherwise the module array stays dead. Change `integer :: ierr = 0` (line 28) to an uninitialised `integer :: ierr` plus an explicit `ierr = 0` assignment, to drop the implicit `SAVE`. Abort with the surface index in the message when `ierr /= 0`. Set `bmin = minval(bstart)`, `bmax = maxval(bstart)` over the whole rank-2 array, and keep `bmod00` as the value from surface 1 (`bmod00` is a single reference normalisation; changing its meaning is out of scope — say so in a comment).
3. `src/samplers.f90:261-287` — in `sample_surface_fieldline_impl`, partition `size(zstart,2)` markers across `n_surf = max(num_surf,1)` deterministically (`ntestpart/n_surf`, remainder to the first surfaces), then per marker draw `xi`, call `binsrc(volstart(:,isurf), 1, npoiper*nper, xi, i)` and read `xstart(:,i,isurf)`. Keep the RNG call order per marker unchanged (`random_number` for the bin, then for the pitch) so `deterministic = .True.` runs with a single surface reproduce today's stream bit-for-bit. `ibins` (`src/params.f90:55`) is a scalar left over from single-surface use; leave its semantics alone or record the last surface index alongside it.
4. `src/sub_alpha_lifetime_can.f90:285` — change `integer, intent(in) :: npoi, ierr` to `integer, intent(in) :: npoi` plus `integer, intent(out) :: ierr`, and set `ierr = 0` before the first `chamb_can` call (line 300). Leave the body otherwise untouched. `src/chamb_m.f90:22` should also gain `intent(out)` on `ierr`.
5. `src/samplers.f90:336` — `sample_grid` builds `xinteg = [sbeg(1), xsize_real*ipart, xsize_real*jpart]`. Either extend it the same way or explicitly reject `num_surf > 1` together with `grid_density > 0`; pick one and state it in `README.md`.
6. `src/simple_main.f90:196` and `:207` — no signature change; confirm the chartmap and VMEC branches both reach the new loop. `needs_bminmax_cache()` (`:942-950`) already returns `.true.` for `num_surf > 1`, so the per-particle `get_bminmax` path in `src/classification.f90:158` picks up automatically.
7. `python/pysimple/__init__.py:275` — the comment there names `xstart`/`volstart` as what `reallocate_arrays` provides for `init_starting_surf`; update it for the new ranks. No new wrapper accessor is needed: `num_surf` is a plain scalar already reachable as `params.num_surf` (`python/pysimple/__init__.py:461`, `:482`), unlike `sbeg`, which needs `params_wrapper.get_sbeg`/`set_sbeg` (`src/params_wrapper.f90:10-20`).
8. `test/tests/test_samplers.f90:50-71` — implement `test_sample_surface` by deleting the `print`/`return` at lines 58-59, then set `num_surf = 2`, `sbeg(1:2) = [0.3d0, 0.7d0]`, call `init_starting_surf` and `sample_surface_fieldline`, and assert both surfaces appear with the expected marker counts.
9. `test/tests/CMakeLists.txt:408-409` — add the missing `add_test(NAME test_samplers COMMAND test_samplers.x)` plus `set_tests_properties(test_samplers PROPERTIES WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR} LABELS "unit")`. This test needs a field, so it likely has to move inside the chartmap-fixture block (lines 509-542) with `DEPENDS "generate_test_boozer_chartmap_data"`, as `test_chartmap_startmode1` does.
10. `test/tests/test_multi_surface_startmode1.py` (new) — model it on `test/tests/test_chartmap_startmode1.py`, which already takes `$<TARGET_FILE:simple.x>` and the chartmap `.nc` as argv and writes its own config. Run `simple.x` twice, with `num_surf = 1, sbeg = 0.3` and `num_surf = 2, sbeg = 0.3, 0.7`, and assert the marker histogram in `start.dat`. Register it next to `test_chartmap_startmode1` with `LABELS "integration;python"` and the same `DEPENDS`.
11. `test/test_data/simple.multiple_fieldline.in` — currently orphaned. Either wire it into the new test or delete it, so the tree stops advertising an untested path. Note its stale comment claims `size(sbeg)==0` distributes in volume, which is not how `num_surf` works.
12. Docs — `README.md` (`sbeg` is discussed at `:289-290`), the repo-root `simple.in:12-13`, and `examples/simple_full.in:13`: state that `num_surf` is the authoritative count of `sbeg` entries for `startmode = 1`, that markers are split evenly, and that `num_surf = 0` means volume sampling while still tracing one line for `bmin`/`bmax`. `DOC/config.md` currently documents none of `sbeg`/`num_surf`/`startmode`; adding a short section there is optional.

No libneo change is required: `integrate_mfl_can`, `chamb_can`, `binsrc` and the sampler all live in SIMPLE.

## Acceptance criteria

- [ ] `make CONFIG=Fast` builds clean with no new warnings.
- [ ] `make test` passes, including the newly registered `test_samplers` and `test_multi_surface_startmode1`.
- [ ] With `num_surf = 2, sbeg = 0.3d0, 0.7d0, ntestpart = 64`, `start.dat` contains 32 markers per surface, each within `1d-6` of its traced field line's radial coordinate.
- [ ] `make test-golden` is unchanged for **all** cases under `test/golden_record/`: `albert`, `albert_coils`, `boozer`, `boozer_ncsx`, `canonical`, `classifier`, `classifier_combined`, `classifier_fast`, `collision`, `meiss`, `meiss_coils`, `test_tokamak`, `test_tokamak_classifier`. All ship a scalar `sbeg`; only `classifier_combined` and `classifier_fast` set `num_surf` at all, both to `0` — those two are the regression risk described above.
- [ ] `startmode = 1` with `num_surf = 0` still produces the same `bmin`/`bmax` and the same single-surface `start.dat` as before the change.
- [ ] `startmode = 1` with `num_surf` greater than `size(sbeg)` stops with a diagnostic naming both keys.
- [ ] `integrate_mfl_can` declares `ierr` as `intent(out)`; a run with `sbeg` outside the chamber reaches the `'starting field line has points outside the chamber'` branch at `src/samplers.f90:42-43`.
- [ ] `params::bstart` is either written by `init_starting_surf` or removed from `src/params.f90`; the local shadow at `src/samplers.f90:29` is gone.

## References

The volume weight built by `integrate_mfl_can` is the flux-tube volume element `dV/d\psi = \oint dl/B` — `dery(3) = 1.d0/(bmod*hctrvr(3))` at `src/sub_alpha_lifetime_can.f90:263`, integrated over `phi`, with `hctrvr(3) = d\phi/dl`. For the flux-surface volume element see D'haeseleer, Hitchon, Callen & Shohet, *Flux Coordinates and Magnetic Field Structure* (Springer, 1991). The SIMPLE method and its marker initialisation are described in Albert, Kasilov & Kernbichler, "Accelerated methods for direct computation of fusion alpha particle losses within stellarator optimization", *J. Plasma Phys.* **86**, 815860201 (2020); the underlying integrators in Albert, Kasilov & Kernbichler, "Symplectic integration with non-canonical quadrature for guiding-center orbits in magnetic confinement devices", *J. Comput. Phys.* **403**, 109065 (2020). Beyond those, this is an implementation detail internal to the code and needs no further literature reference.

## Scope and risk

Single PR, roughly `size/M`: the array-rank change in `src/params.f90` propagates into `src/samplers.f90` and a Python wrapper comment, and every consumer of `xstart`/`volstart`/`bstart` is listed above (`rg -n 'bstart|volstart|xstart' src/ test/` is the full set). Two risks dominate: (1) the `num_surf = 0` golden cases, mitigated by the `max(num_surf,1)` rule in step 2 and gated by `make test-golden`; (2) perturbing the random-number stream and thereby the golden records, mitigated by step 3's ordering constraint. Related but distinct: itpplasma/SIMPLE#359 changes what `sbeg` *means* radially on the scaled REFCOORDS path and touches the same `init_starting_surf` entry point, so coordinate the two PRs and rebase whichever lands second.
